r2dbc-spi/src/main/java/io/r2dbc/spi/ColumnMetadata.java

```diff
@@ -31,17 +30,102 @@
+     *
+     * @return the precision of the column or {@code null} if the precision is not available.
+     */
+    @Nullable
+    default Integer getPrecision() {
+        return null;
+    }
+
+    /**
+     * Returns the scale of the column.
+     * <p>
+     * This is the number of digits to right of the decimal point.
+     * Returns {@code null} for data types where the scale is not applicable  or if the precision cannot be provided.
+     * <p>
+     * <strong>Implementation notes</strong>
+     * Implementation of this method is optional. The default implementation returns {@code null}.
      *
-     * @return the precision of the column
+     * @return the scale of the column or {@code null} if the scale is not available.
      */
-    Optional<Integer> getPrecision();
+    @Nullable
+    default Integer getScale() {
+        return null;
+    }
 
     /**
-     * Returns the type of the column.
+     * Returns the nullability of column values.
+     * <p>
+     * <strong>Implementation notes</strong>
+     * Implementation of this method is optional. The default implementation returns {@link Nullability#Unknown}.
      *
-     * @return the type of the column
+     * @return the nullability of column values.
+     * @see Nullability
      */
-    Integer getType();
+    default Nullability getNullability() {
+        return Nullability.Unknown;
+    }
+
+    /**
+     * Returns the primary Java {@link Class type}. This type can be considered the native representation that is used to exchange values with the least loss in precision.
+     * <p>
+     * <strong>Implementation notes</strong>
+     * Drivers should implement this method. The method should return the actual type and refrain from returning {@code Object.class}. The default implementation returns {@code null}.
+     * Drivers may need to inspect the value and perform parts of decoding to determine the native Java type. This method
+     * should be expected to run in non-constant time.
+     *
+     * @return the primary Java {@link Class type} or {@code null} if the type is not available.
+     * @see Row#get(Object, Class)
+     */
+    @Nullable
+    default Class<?> getJavaType() {
+        return null;
+    }
+
+    /**
+     * Returns the native type descriptor that potentially exposes more metadata.
+     * <p>
+     * <strong>Implementation notes</strong>
+     * Drivers should implement this method if they can expose a driver-specific type metadata object exposing additional information. The default implementation returns {@code null}.
+     *
+     * @return the native type descriptor that potentially exposes more metadata or {@code null} if no native type descriptor is available.
+     */
+    @Nullable
+    default Object getNativeTypeMetadata() {
+        return null;
+    }
+
+    /**
+     * Constants indicating nullability of column values.
+     */
+    enum Nullability {
```

Comment:
Let's promote this to a top-level type.  I realize that it is scoped to this class, but given how small the API is and that we've promoted all the rest of the `enum`s up I think we should maintain consistency.

Reply:
Done.